codec/grpc/grpc_codec.go

```diff
@@ -1,33 +1,65 @@
 package grpc
 
-import "fmt"
+import (
+	"google.golang.org/grpc/encoding"
+	"google.golang.org/grpc/mem"
+
+	// Guarantee that the built-in proto is called registered before this one
+	// so that it can be replaced.
+	_ "google.golang.org/grpc/encoding/proto"
+)
 
 // Name is the name registered for the proto compressor.
 const Name = "proto"
 
-type Codec struct{}
-
 type vtprotoMessage interface {
-	MarshalVT() ([]byte, error)
+	MarshalToVT(data []byte) (int, error)
 	UnmarshalVT([]byte) error
+	SizeVT() int
+}
+
+type Codec struct {
+	fallback encoding.CodecV2
 }
 
-func (Codec) Marshal(v interface{}) ([]byte, error) {
-	vt, ok := v.(vtprotoMessage)
-	if !ok {
-		return nil, fmt.Errorf("failed to marshal, message is %T (missing vtprotobuf helpers)", v)
+func (Codec) Name() string { return Name }
+
+func (c *Codec) Marshal(v any) (data mem.BufferSlice, err error) {
+	if m, ok := v.(vtprotoMessage); ok {
+		size := m.SizeVT()
+		if mem.IsBelowBufferPoolingThreshold(size) {
+			buf := make([]byte, 0, size)
+			if _, err := m.MarshalToVT(buf[:0]); err != nil {
+				return nil, err
+			}
+			data = append(data, mem.SliceBuffer(buf))
+		} else {
+			pool := mem.DefaultBufferPool()
+			buf := pool.Get(size)
+			if _, err := m.MarshalToVT((*buf)[:0]); err != nil {
+				pool.Put(buf)
+				return nil, err
+			}
+			data = append(data, mem.NewBuffer(buf, pool))
```

Comment:
I haven't tested this yet, but similarly to the case above, I expect something like this would work better:
```suggestion
			if _, err := m.MarshalToSizedBufferVT((*buf)[:size]); err != nil {
				pool.Put(buf)
				return nil, err
			}
			data = mem.BufferSlice{mem.NewBuffer(buf, pool)}
```